kong/plugins/rate-limiting/handler.lua

```diff
@@ -182,38 +173,23 @@ function RateLimitingHandler:access(conf)
       headers[RATELIMIT_LIMIT] = limit
       headers[RATELIMIT_REMAINING] = remaining
       headers[RATELIMIT_RESET] = reset
-
-      kong.ctx.plugin.headers = headers
     end
 
     -- If limit is exceeded, terminate the request
     if stop then
-      return kong.response.error(429, "API rate limit exceeded", {
-        [RETRY_AFTER] = reset
-      })
+      headers = headers or {}
+      headers[RETRY_AFTER] = reset
+      return kong.response.error(429, "API rate limit exceeded", headers)
     end
-  end
 
-  kong.ctx.plugin.timer = function()
-    local ok, err = timer_at(0, increment, conf, limits, identifier, current_timestamp, 1)
-    if not ok then
-      kong.log.err("failed to create timer: ", err)
+    if headers then
+      kong.response.set_headers(headers)
     end
   end
-end
-
 
-function RateLimitingHandler:header_filter(_)
-  local headers = kong.ctx.plugin.headers
-  if headers then
-    kong.response.set_headers(headers)
-  end
-end
-
-
-function RateLimitingHandler:log(_)
-  if kong.ctx.plugin.timer then
-    kong.ctx.plugin.timer()
+  local _, err = policies[conf.policy].increment(conf, limits, identifier, current_timestamp, 1)
```

Comment:
I agree with removing the timer, but wouldn't it make more sense to keep the increment in the log phase?

As I understand it, putting the increment in `access` will hold the request until after the increment is done (possibly several database access, if we're using cluster policy). The log phase happens when the response has already been sent back to the client so it shouldn't be penalized in that manner.